validate.go

```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+	`[pP]ingdom`:            {`pingdom`, `Pingdom`},
+	`Bark[rR]owler`:         {`Barkrowler`, `BarkRowler`},
+	`^Apache-HttpClient`:    {`^Apache-HttpClient`},
+	`^LCC `:                 {`^LCC `},
+	`(^| )sentry\/`:         {`^sentry/`, ` sentry/`},
+	`^curl`:                 {`^curl`},
+	`[Cc]urebot`:            {`Curebot`, `curebot`},
+	`^PHP-Curl-Class`:       {`^PHP-Curl-Class`},
+	`(^| )PTST\/`:           {`^PTST/`, ` PTST/`},
+	`^BW\/`:                 {`^BW/`},
+}
+
+var pattern2mainLiteral = map[string]string{
+	`AdsBot-Google([^-]|$)`:             `AdsBot-Google`,
+	`BlogTraffic\/\d\.\d+ Feed-Fetcher`: `BlogTraffic/`,
+}
+
+func analyzePattern(pattern string) (olds []string, re *regexp.Regexp) {
+	literals, has := pattern2literals[pattern]
+	if has {
+		return literals, nil
+	}
+
+	re = regexp.MustCompile(pattern)
+	prefix, complete := re.LiteralPrefix()
+	if complete {
+		return []string{prefix}, nil
+	}
+
+	mainLiternal, has := pattern2mainLiteral[pattern]
+	if !has {
+		panic("don't know what to do with pattern: " + pattern)
+	}
+	return []string{mainLiternal}, re
+}
+
+type regexpPattern struct {
+	re    *regexp.Regexp
+	index int
+}
+
+type matcher struct {
+	replacer *strings.Replacer
+	regexps  []regexpPattern
+}
+
+var uniqueToken = hex.EncodeToString((&maphash.Hash{}).Sum(nil))
+
+const (
+	literalLabel = '-'
+	regexpLabel  = '*'
+)
+
+var m = func() matcher {
+	regexps := []regexpPattern{}
+	oldnew := make([]string, 0, len(Crawlers)*2)
+
+	// Put re-based patterns to the end to prevent AdsBot-Google from
+	// shadowing AdsBot-Google-Mobile.
+	var oldnew2 []string
+
 	for i, crawler := range Crawlers {
-		regexps[i] = regexp.MustCompile(crawler.Pattern)
+		literals, re := analyzePattern(crawler.Pattern)
+
+		label := literalLabel
+		num := i
+		if re != nil {
+			label = regexpLabel
+			num = len(regexps)
+			regexps = append(regexps, regexpPattern{
+				re:    re,
+				index: i,
+			})
+		}
+
+		replaceWith := fmt.Sprintf(" %s%c%05d ", uniqueToken, label, num)
+
+		for _, literal := range literals {
+			if re != nil {
+				oldnew2 = append(oldnew2, literal, replaceWith)
+			} else {
+				oldnew = append(oldnew, literal, replaceWith)
+			}
+		}
+	}
+	oldnew = append(oldnew, oldnew2...)
+
+	regexps2 := make([]regexpPattern, len(regexps))
+	copy(regexps2, regexps)
+
+	r := strings.NewReplacer(oldnew...)
+	r.Replace("") // To cause internal build process.
+
+	return matcher{
+		replacer: r,
+		regexps:  regexps2,
 	}
-	return regexps
 }()
 
 // Returns if User Agent string matches any of crawler patterns.
 func IsCrawler(userAgent string) bool {
-	for _, re := range regexps {
-		if re.MatchString(userAgent) {
+	// This code is mostly copy-paste of MatchingCrawlers,
+	// but with early exit logic, so it works a but faster.
+
+	text := "^" + userAgent + "$"
+	replaced := m.replacer.Replace(text)
+	if replaced == text {
+		return false
+	}
+
+	for {
+		uniquePos := strings.Index(replaced, uniqueToken)
+		if uniquePos == -1 {
```

Comment:
Can you elaborate on this, please?

```go
for uniquePos := strings.Index(replaced, uniqueToken); uniquePos != -1; uniquePos = strings.Index(replaced, uniqueToken) {
}
```

Like this? I don't really like it, because it copy-pastes the `strings.Index` call.